### PR TITLE
ocr3 - evm commit report encoder

### PR DESCRIPTION
**Pull request by dimkouv**
Implementation of evm commit report encoder.

### CODE REVIEW

File: core/services/ocr3/plugins/ccipevm/commitcodec.go [hunk 1]
```diff
@@ -2,22 +2,156 @@ package ccipevm
...
+	"github.com/smartcontractkit/chainlink/v2/core/services/ocr2/plugins/ccip/abihelpers"
 )
 
-var _ cciptypes.CommitPluginCodec = (*CommitPluginCodec)(nil)
-
-type CommitPluginCodec struct{}
+// CommitPluginCodecV1 is a codec for encoding and decoding commit plugin reports.
+// Compatible with:
+// - "EVM2EVMMultiOffRamp 1.6.0-dev"
+type CommitPluginCodecV1 struct {
```
Comment (makramkd) [inline core/services/ocr3/plugins/ccipevm/commitcodec.go:21]: I wonder if this should be in a folder `v1_6_0` to follow the convention we have in OCR2 folder

Comment (makramkd) [reply core/services/ocr3/plugins/ccipevm/commitcodec.go:21]: Ditto for the msg hasher

File: core/services/ocr3/plugins/ccipevm/commitcodec.go [hunk 2]
```diff
@@ -2,22 +2,156 @@ package ccipevm
...
+		} `json:"priceUpdates"`
+		MerkleRoots []struct {
+			SourceChainSelector uint64 `json:"sourceChainSelector"`
+			Interval            struct {
+				Min uint64 `json:"min"`
+				Max uint64 `json:"max"`
+			} `json:"interval"`
+			MerkleRoot [32]uint8 `json:"merkleRoot"`
+		} `json:"merkleRoots"`
+	})
```
Comment (makramkd) [inline core/services/ocr3/plugins/ccipevm/commitcodec.go]: Why this anonymous struct?

Comment (makramkd) [reply core/services/ocr3/plugins/ccipevm/commitcodec.go]: Shouldn't this use the gethwrapper type?

File: core/services/ocr3/plugins/ccipevm/commitcodec.go [hunk 3]
```diff
@@ -2,22 +2,156 @@ package ccipevm
...
+
+	if !is {
+		return cciptypes.CommitPluginReport{},
+			fmt.Errorf("expected EVM2EVMMultiOffRampCommitReport, got %T", unpacked[0])
+	}
+
+	merkleRoots := make([]cciptypes.MerkleRootChain, 0, len(commitReport.MerkleRoots))
```
Comment (0xnogo) [inline core/services/ocr3/plugins/ccipevm/commitcodec.go:100]: I am wondering if creating helper functions will help readability. It would give something like this:
```
	merkleRoots := convertMerkleRoots(commitReport.MerkleRoots)
	tokenPriceUpdates := convertTokenPriceUpdates(commitReport.PriceUpdates.TokenPriceUpdates)
	gasPriceUpdates := convertGasPriceUpdates(commitReport.PriceUpdates.GasPriceUpdates)

	return cciptypes.CommitPluginReport{
		MerkleRoots: merkleRoots,
		PriceUpdates: cciptypes.PriceUpdates{
			TokenPriceUpdates: tokenPriceUpdates,
			GasPriceUpdates:   gasPriceUpdates,
		},
	}, nil
```

No strong opinion on this one.

File: core/services/ocr3/plugins/ccipevm/commitcodec_test.go [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+		},
+		GasPriceUpdates: []cciptypes.GasPriceChain{
+			{GasPrice: cciptypes.NewBigInt(utils.RandUint256()), ChainSel: cciptypes.ChainSelector(rand.Uint64())},
+			{GasPrice: cciptypes.NewBigInt(utils.RandUint256()), ChainSel: cciptypes.ChainSelector(rand.Uint64())},
+			{GasPrice: cciptypes.NewBigInt(utils.RandUint256()), ChainSel: cciptypes.ChainSelector(rand.Uint64())},
+		},
+	},
+}
+
+func TestCommitPluginCodec(t *testing.T) {
```
Comment (0xnogo) [inline core/services/ocr3/plugins/ccipevm/commitcodec_test.go]: We might be missing the error handling cases (test `Encode` method with an invalid token address, an empty token price, and an empty gas price)

File: core/services/ocr3/plugins/ccipevm/commitcodec.go (hunk 2 above)
Comment (dimkouv) [reply core/services/ocr3/plugins/ccipevm/commitcodec.go]: doesn't work

File: core/services/ocr3/plugins/ccipevm/commitcodec.go (hunk 1 above)
Comment (dimkouv) [reply core/services/ocr3/plugins/ccipevm/commitcodec.go:21]: my idea was:
```
msgHasherV1 supports:
   onRamp1.6.0
   onRamp1.7.0
   onRamp1.8.0

msgHasherV2 supports:
      onRamp1.9.0
     ...
```

File: core/services/ocr3/plugins/ccipevm/commitcodec.go (hunk 2 above)
Comment (makramkd) [reply core/services/ocr3/plugins/ccipevm/commitcodec.go]: Can you try this instead?

```golang
out := *abi.ConvertType(unpacked[0], new(evm_2_evm_multi_offramp.EVM2EVMMultiOffRampCommitReport)).(*evm_2_evm_multi_offramp.EVM2EVMMultiOffRampCommitReport)
```

This is what the gethwrappers do.

Comment (dimkouv) [reply core/services/ocr3/plugins/ccipevm/commitcodec.go]: awesome! thanks

File: core/services/ocr3/plugins/ccipevm/commitcodec.go (hunk 3 above)
Comment (dimkouv) [reply core/services/ocr3/plugins/ccipevm/commitcodec.go:100]: let's do this if we require it more than once

File: core/services/ocr3/plugins/ccipevm/commitcodec_test.go (hunk 4 above)
Comment (dimkouv) [reply core/services/ocr3/plugins/ccipevm/commitcodec_test.go]: nice, added

File: core/services/ocr3/plugins/ccipevm/commitcodec.go [hunk 5]
```diff
@@ -2,22 +2,137 @@ package ccipevm
...
+			return nil, fmt.Errorf("empty gas price for chain: %d", update.ChainSel)
+		}
+
+		gasPriceUpdates = append(gasPriceUpdates, evm_2_evm_multi_offramp.InternalGasPriceUpdate{
+			DestChainSelector: uint64(update.ChainSel),
+			UsdPerUnitGas:     update.GasPrice.Int,
+		})
+	}
+
+	evmReport := evm_2_evm_multi_offramp.EVM2EVMMultiOffRampCommitReport{
```
Comment (asoliman92) [inline core/services/ocr3/plugins/ccipevm/commitcodec.go:73]: Is there a possibility for adding a version to the report? Is there a case where we might need to change the en/decoding of the report with same offRamp version?

**Review by asoliman92 — APPROVED** [review]
LGTM

File: core/services/ocr3/plugins/ccipevm/commitcodec.go (hunk 5 above)
Comment (dimkouv) [reply core/services/ocr3/plugins/ccipevm/commitcodec.go:73]: for the same offRamp no.
for a new offRamp (with breaking  changes) version yes, in that case we create a `v2`

**Review by makramkd — APPROVED** [review]

**Review by 0xnogo — APPROVED** [review]
